### PR TITLE
Add property propagation in memo optimization

### PR DESCRIPTION
**Pull request by arzuschen**
This patch attempts to add a new property propagation step to the end the exploration. The groups are traversed from top to down and properties required by certain physical nodes are propagated to the child group. If a member in the child group can provided the property, it will be added to the candidate list.

The locate min cost member functions are changed to accommodate to property requirements. If the property requirement is null, then the min cost member is found and returned same as before. When a property requirement is present, the min cost member with additional enforcement node will be compared to all the other members that can provide that property. Then the optimal member will be found.

Since enforced node will be added onto the member physical node, the assumption that each member correspond to one layer of physical node is no longer valid. However, since bypass is created for join solver case, which will also work for enforced node case, minimal changes is required for copy out min logic physic plan function.

### CODE REVIEW

File: qpmodel/optimizer.cs [hunk 1]
```diff
@@ -50,8 +51,67 @@ public class Property { }
     public class PhysicProperty : Property
     {
         // ordering: the ordered expression and whether is descending
-        KeyValuePair<Expr, bool> ordering_;
-        KeyValuePair<Expr, DistributionType> distribution_;
+        List<KeyValuePair<Expr, bool>> ordering_;
+        List<KeyValuePair<Expr, DistributionType>> distribution_;
+        public PhysicProperty(List<Expr> order, List<bool> desc = null)
```
Comment (zhouqingqing) [inline qpmodel/optimizer.cs:56]: https://coveralls.io/builds/31922927/source?filename=qpmodel/optimizer.cs

Seems these code are not exercised in unittest, can you add them?

Comment (arzuschen) [reply qpmodel/optimizer.cs:56]: Sure

Comment (zhouqingqing) [reply qpmodel/optimizer.cs:56]: also, consider removing distribution_ and introducing it in your next PR with unittest.

File: qpmodel/PhysicalNode.cs [hunk 2]
```diff
@@ -1057,7 +1060,18 @@ public class PhysicStreamAgg : PhysicAgg
...
+        {
+            var exprlist = (logic_ as LogicAgg).groupby_;
+            if (exprlist is null) return null;
+            return new PhysicProperty(exprlist);
+        }
+        public override PhysicProperty SuppiedProperty()
+        {
+            var exprlist = (logic_ as LogicAgg).groupby_;
+            if (exprlist is null) return null;
+            return new PhysicProperty(exprlist);
```
Comment (zhouqingqing) [inline qpmodel/PhysicalNode.cs:1073]: do you mean it supplies (after streamAgg) sorted attributes on group by keys? If so, return "PhysicProperty" is not clear in naming. You may want to derive a "SortOrderProperty" out of Property class for this purpose?

File: qpmodel/optimizer.cs [hunk 3]
```diff
@@ -35,6 +35,7 @@
 using qpmodel.expr;
 
 using LogicSignature = System.Int64;
+using qpmodel.utils;
```
Comment (zhouqingqing) [inline qpmodel/optimizer.cs:38]: move to L36

File: qpmodel/optimizer.cs [hunk 4]
```diff
@@ -50,8 +51,61 @@ public class Property { }
     public class PhysicProperty : Property
```
Comment (zhouqingqing) [inline qpmodel/optimizer.cs:51]: See previous comment, you may want to have SortOrderProperty/DistributionProperty etc derived from Property? For clear naming.

File: qpmodel/optimizer.cs [hunk 5]
```diff
@@ -50,8 +51,61 @@ public class Property { }
...
     {
         // ordering: the ordered expression and whether is descending
-        KeyValuePair<Expr, bool> ordering_;
-        KeyValuePair<Expr, DistributionType> distribution_;
+        List<KeyValuePair<Expr, bool>> ordering_;
```
Comment (zhouqingqing) [inline qpmodel/optimizer.cs:54]: We may have COLLATE in future in sort clause, so sort is ORDER BY <expr> <desc?> <collation?>. You can have a refactoring patch to introduce a SortByExpr class grouping them together.

File: qpmodel/optimizer.cs [hunk 6]
```diff
@@ -50,8 +51,61 @@ public class Property { }
...
+            for (int i = 0; i < ordering_.Count; i++)
+            {
+                if (!ordering_[i].Key.Equals(other.ordering_[i].Key)) return false;
+                if (ordering_[i].Value != other.ordering_[i].Value) return false;
+            }
+            return true;
+        }
+        public override bool Equals(object obj)
+        {
+            if (obj.GetType() != this.GetType()) return false;
```
Comment (zhouqingqing) [inline qpmodel/optimizer.cs:102]: if (!(obj is PhysicProperty)) return false?

File: qpmodel/optimizer.cs [hunk 7]
```diff
@@ -50,8 +51,61 @@ public class Property { }
...
+                order.Add(pair.Key);
+                desc.Add(pair.Value);
+            }
+            var logicnode = new LogicOrder(node.logic_, order, desc);
+            return new PhysicOrder(logicnode, node);
+        }
+
+        public bool Equals(PhysicProperty other)
+        {
+            if (other is null) return false;
```
Comment (zhouqingqing) [inline qpmodel/optimizer.cs:91]: no null checks here since you already did it in caller.

File: qpmodel/optimizer.cs (hunk 6 above)
Comment (arzuschen) [reply qpmodel/optimizer.cs:102]: I wanted to declare that if if is compared to a object that is not PhysicProperty then they cannot be equal.

File: qpmodel/PhysicalNode.cs [hunk 8]
```diff
@@ -1099,7 +1113,7 @@ public override string Exec(Func<Row, string> callback)
                     {
                         // output current grouped row if any
                         if (curGroupKey != null)
-                            FinalizeAGroupRow(context, keys, curGroupRow, callback);
+                            FinalizeAGroupRow(context, curGroupKey, curGroupRow, callback);
```
Comment (zhouqingqing) [inline qpmodel/PhysicalNode.cs:1116]: please send in a separate PR for this bug fix - I will do a quick merge.

File: qpmodel/optimizer.cs [hunk 9]
```diff
@@ -345,11 +405,44 @@ public void ExploreGroup(Memo memo)
...
+                propertyCandidates_.Add(property, new List<int>());
+            if (property is null)
+            {
+                if (propagated_) return;
+                propagated_ = true;
+            }
+
+            for (int i = 0; i < exprList_.Count; i++)
+            {
+                if (exprList_[i].physic_ != null && !solveroptimized)
```
Comment (zhouqingqing) [inline qpmodel/optimizer.cs:427]: is it possible to follow ExplorGroup() style like:
  EnforceGroup()
     foreach (member in exprList_)
         EnforceMember();

File: qpmodel/optimizer.cs [hunk 10]
```diff
@@ -713,7 +882,7 @@ public PhysicNode CopyOutOnePlan(SQLStatement stmt, Memo memo)
 
             // retrieve the lowest cost plan
             Debug.Assert(stmt.physicPlan_ is null);
-            stmt.physicPlan_ = memo.rootgroup_.CopyOutMinLogicPhysicPlan();
+            stmt.physicPlan_ = memo.rootgroup_.CopyOutMinLogicPhysicPlan(memo.baseproperty_);
             stmt.logicPlan_ = stmt.physicPlan_.logic_;
```
Comment (zhouqingqing) [inline qpmodel/optimizer.cs:887]: please print memo for better illustration

File: qpmodel/optimizer.cs (hunk 7 above)
Comment (arzuschen) [reply qpmodel/optimizer.cs:91]: I think I still need this because the caller only checks for type

**Pull request closed without merging**